### PR TITLE
fix height hint Zero issue in utxonursery

### PR DESCRIPTION
**Pull request by ziggie1984**
Fixes: https://github.com/lightningnetwork/lnd/issues/8383

please try-out this patch if you can @robtex

### CODE REVIEW

**Comment by ziggie1984** [comment]
That was the error for the reviewer: 

```
2025-10-06 14:51:37.587 [INF] UTXN: Publishing CLTV-delayed HTLC output using timeout tx (txid=62ff2f1f063082090be8aecd42bba5ae29b4b1b1a00e1f13806b9d99f8087e00): (*wire.MsgTx)(0xc007f7fc00)({
 Version: (int32) 2,
 TxIn: ([]*wire.TxIn) (len=1 cap=1) {
  (*wire.TxIn)(0xc003b4ccc0)({
   PreviousOutPoint: (wire.OutPoint) 81e84010f3eb56e10eb808b00711c46cf1a96368015f83a21223c979c7b846f9:1,
   SignatureScript: ([]uint8) {
   },
   Witness: (wire.TxWitness) (len=5 cap=5) {
    ([]uint8) {
    },
    ([]uint8) (len=72 cap=72) {
     00000000  30 45 02 21 00 87 d8 f4  39 83 5a 5d 71 46 17 fb  |0E.!....9.Z]qF..|
     00000010  b4 c1 5b aa fd 91 89 9d  5c 30 82 fe 58 61 94 6c  |..[.....\0..Xa.l|
     00000020  dd 58 af b3 38 02 20 19  33 74 07 a1 b8 c8 6d 43  |.X..8. .3t....mC|
     00000030  a1 49 24 e9 a0 08 8e 4e  d7 e0 94 de d3 99 e6 40  |.I$....N.......@|
     00000040  67 98 f4 92 43 27 92 01                           |g...C'..|
    },
    ([]uint8) (len=71 cap=71) {
     00000000  30 44 02 20 0d b3 01 32  c5 c3 b9 97 a2 bd 66 a1  |0D. ...2......f.|
     00000010  aa c5 2a 55 6a 42 1f 2d  30 4d 6f 51 0a 66 b6 d7  |..*UjB.-0MoQ.f..|
     00000020  0d ae 4d 57 02 20 44 d0  f9 62 38 81 f2 8d be f4  |..MW. D..b8.....|
     00000030  57 48 7e 0a 36 dd 7f 0c  30 fb c9 1c a4 9e ba 25  |WH~.6...0......%|
     00000040  64 25 04 7b 4d 39 01                              |d%.{M9.|
    },
    ([]uint8) {
    },
    ([]uint8) (len=133 cap=133) {
     00000000  76 a9 14 69 d9 29 69 40  47 52 24 bd 20 64 c3 52  |v..i.)i@GR$. d.R|
     00000010  10 f7 44 08 b5 18 72 87  63 ac 67 21 03 41 0a f5  |..D...r.c.g!.A..|
     00000020  bb e6 aa b1 37 6b 79 26  df 99 3d 7f 8d a7 6f 39  |....7ky&..=...o9|
     00000030  90 4c 1c 21 02 59 fe 61  72 72 6c 66 4c 7c 82 01  |.L.!.Y.arrlfL|..|
     00000040  20 87 64 75 52 7c 21 02  9c 84 44 ae a5 0f 39 df  | .duR|!...D...9.|
     00000050  08 19 f6 99 d9 72 36 ee  b5 53 ab c6 e8 8d a6 32  |.....r6..S.....2|
     00000060  5f 0e 2d 14 4a 1a 0f 3b  52 ae 67 a9 14 b9 33 f0  |_.-.J..;R.g...3.|
     00000070  45 99 7c 8c 49 52 af 17  92 c8 a4 bd 01 5a 29 be  |E.|.IR.......Z).|
     00000080  c4 88 ac 68 68                                    |...hh|
    }
   },
   Sequence: (uint32) 0
  })
 },
 TxOut: ([]*wire.TxOut) (len=1 cap=1) {
  (*wire.TxOut)(0xc007f275a0)({
   Value: (int64) 15072,
   PkScript: ([]uint8) (len=34 cap=34) {
    00000000  00 20 e2 f3 5d 12 f8 c8  d0 14 cd bb ab 65 65 c8  |. ..]........ee.|
    00000010  12 05 67 e2 aa bd eb 89  ef 05 bb c7 9e 89 e9 d5  |..g.............|
    00000020  1e 9d                                             |..|
   }
  })
 },
 LockTime: (uint32) 0
})

2025-10-06 14:51:37.730 [ERR] BTCN: Broadcast attempt failed: Transaction already in block chain
2025-10-06 14:51:37.862 [WRN] BTWL: Transaction 62ff2f1f063082090be8aecd42bba5ae29b4b1b1a00e1f13806b9d99f8087e00 not accepted by mempool: txn-already-known
2025-10-06 14:51:37.862 [INF] BTWL: Inserting unconfirmed transaction 62ff2f1f063082090be8aecd42bba5ae29b4b1b1a00e1f13806b9d99f8087e00
2025-10-06 14:51:37.998 [INF] BTWL: 62ff2f1f063082090be8aecd42bba5ae29b4b1b1a00e1f13806b9d99f8087e00: tx already confirmed
2025-10-06 14:51:38.130 [ERR] UTXN: Failed to sweep first-stage HTLC (CLTV-delayed) output 62ff2f1f063082090be8aecd42bba5ae29b4b1b1a00e1f13806b9d99f8087e00:0
2025-10-06 14:51:38.130 [ERR] UTXN: Failed to sweep outputs at height=0: a height hint greater than 0 must be provided
2025-10-06 14:51:38.130 [ERR] UTXN: Failed to reload class: a height hint greater than 0 must be provided
```

**Comment by ziggie1984** [comment]
The patch worked as foreseen.

The node operator had however one node which had corrupted data entries, if this happens we error out which is the best way imo otherwise we do not recognise this behaviour.


```
2025-10-07 16:42:53.098 [ERR] UTXN: Failed to sweep first-stage HTLC (CLTV-delayed) output 1ce1ca1ebcf883ce95cd134291d510710447c477daadc6de90b5eb4155170b20:620756992
2025-10-07 16:42:53.098 [ERR] UTXN: Failed to sweep outputs at height=0: cannot determine height hint for crib output with expiry=0: unable to find closed channel summary
2025-10-07 16:42:53.098 [ERR] UTXN: Failed to reload class: cannot determine height hint for crib output with expiry=0: unable to find closed channel summary
2025-10-07 16:42:53.098 [INF] UTXN: UTXO nursery shutting down...
```

We removed those entries manually, they can still be swept with chantools.

File: contractcourt/utxonursery.go [hunk 1]
```diff
@@ -242,6 +242,57 @@ func NewUtxoNursery(cfg *NurseryConfig) *UtxoNursery {
...
+	// Use the close height minus the confirmation depth as a
+	// conservative height hint. This ensures we don't miss the
+	// confirmation even if it happened around the close height.
+	var heightHint uint32
+	if closeSummary.CloseHeight > u.cfg.ConfDepth {
+		heightHint = closeSummary.CloseHeight - u.cfg.ConfDepth
+	} else {
+		// If close height is very low, use 1 as the minimum
+		// valid height hint.
+		heightHint = 1
```
Comment (Roasbeef) [inline contractcourt/utxonursery.go]: Rather than 1, we shoudl use height `481824` as that's the segwit activation height. No channels existed before that point.

File: contractcourt/utxonursery_test.go [hunk 2]
```diff
@@ -1262,3 +1262,150 @@ func TestKidOutputDecode(t *testing.T) {
...
+					}, nil
+				},
+			}
+
+			// Create nursery instance.
+			nursery := &UtxoNursery{
+				cfg: cfg,
+			}
+
+			// Call the function under test.
```
Comment (Roasbeef) [inline contractcourt/utxonursery_test.go]: Superflous comments here and above in the test.

File: contractcourt/utxonursery.go (hunk 1 above)
Comment (Roasbeef) [reply contractcourt/utxonursery.go]: Do we need to handle the case where `CloseHeight` is actually zero? In which case, we can use the block height in the scid as a lower bound.

Comment (ziggie1984) [reply contractcourt/utxonursery.go]: good idea will use it for this case rather than going for the segwit height

Comment (ziggie1984) [reply contractcourt/utxonursery.go]: ok I added more safety checks and will actually error out if the `closeHeight` and `ShortChannelID` is not set. This makes the software more robust I would say. Let's go with that one.

File: contractcourt/utxonursery_test.go (hunk 2 above)
Comment (ziggie1984) [reply contractcourt/utxonursery_test.go]: removed.

File: contractcourt/utxonursery.go [hunk 3]
```diff
@@ -242,6 +242,71 @@ func NewUtxoNursery(cfg *NurseryConfig) *UtxoNursery {
...
+		heightHint = closeSummary.ShortChanID.BlockHeight
+	}
+
+	// This should never happen the segwit activation height was 481824
+	// so the close height should never be smaller than the conf depth.
+	if heightHint < u.cfg.ConfDepth {
+		return 0, fmt.Errorf("cannot determine height hint for " +
+			"crib output with expiry=0: close height is smaller " +
+			"than conf depth")
+	}
```
[the remark this answers is not in the document]
Comment (ziggie1984) [reply contractcourt/utxonursery.go]: @gemini-code-assist  we check before that the height_hint is not zero by making sure teh closeheight nor the channel scid blockheight are. not zero

**Comment by ziggie1984** [comment]
@gemini-code-assist can you check if this current approach solves all the priority problems from before.

**Comment by ziggie1984** [comment]
Decided to still error out in case the closeSummary is borked.

**Comment by robtex** [comment]
> Fixes: #8383
> 
> please try-out this patch if you can @robtex

Confirmed this helped two really old nodes up and running again. 
Both nodes have had some patches before.

Thanks a lot!!!

**Comment by Roasbeef** [comment]
Tests need to be updated after the latest change.

File: contractcourt/utxonursery.go [hunk 4]
```diff
@@ -163,6 +163,10 @@ const (
 	// kgtnOutputConfTarget is the default confirmation target we'll use for
 	// sweeps of CSV delayed outputs.
 	kgtnOutputConfTarget = 6
+
+	// segwitActiveHeight is the height at which segwit was activated. There
+	// can only be lightning channels after segwit was activated.
+	segwitActiveHeight = 481824
```
[the remark this answers is not in the document]
Comment (ziggie1984) [reply contractcourt/utxonursery.go]: decided to remove this check to have a cleaner design rather than hardcoding mainnet values, this should never happen in the first place anyways.

File: contractcourt/utxonursery_test.go [hunk 5]
```diff
@@ -1262,3 +1263,197 @@ func TestKidOutputDecode(t *testing.T) {
 		})
 	}
 }
+
+// TestPatchZeroHeightHint tests the patchZeroHeightHint function to ensure
+// it correctly handles both normal cases and the edge case where classHeight
+// is zero due to a historical bug.
+func TestPatchZeroHeightHint(t *testing.T) {
```
Comment (yyforyongyu) [inline contractcourt/utxonursery_test.go:1270]: 👍

File: contractcourt/utxonursery.go [hunk 6]
```diff
@@ -242,6 +242,71 @@ func NewUtxoNursery(cfg *NurseryConfig) *UtxoNursery {
...
+			return 0, fmt.Errorf("cannot use fallback height " +
+				"hint: close height is 0 and short " +
+				"channel ID block height is 0")
+		}
+
+		heightHint = closeSummary.ShortChanID.BlockHeight
+	}
+
+	// At this point the height hint should normally be greater than the
+	// conf. depth since channels should have a minimum close height of the
```
Comment (yyforyongyu) [inline contractcourt/utxonursery.go]:
```suggestion
	// conf depth since channels should have a minimum close height of the
```

File: contractcourt/utxonursery.go [hunk 7]
```diff
@@ -242,6 +242,71 @@ func NewUtxoNursery(cfg *NurseryConfig) *UtxoNursery {
...
+	// segwit activation height and the conf. depth which is a config
```
Comment (yyforyongyu) [inline contractcourt/utxonursery.go]:
```suggestion
	// segwit activation height and the conf depth which is a config
```

File: contractcourt/utxonursery.go [hunk 8]
```diff
@@ -242,6 +242,71 @@ func NewUtxoNursery(cfg *NurseryConfig) *UtxoNursery {
...
+	// parameter should be in the single digit range.
+	if heightHint <= u.cfg.ConfDepth {
```
Comment (yyforyongyu) [inline contractcourt/utxonursery.go:292]: Normally this should be impossible right?

File: contractcourt/utxonursery.go [hunk 9]
```diff
@@ -242,6 +242,71 @@ func NewUtxoNursery(cfg *NurseryConfig) *UtxoNursery {
...
+		return 0, fmt.Errorf("cannot use fallback height hint: "+
+			"fallback height hint %v <= confirmation depth %v",
+			heightHint, u.cfg.ConfDepth)
+	}
+
+	// Use the close height minus the confirmation depth as a conservative
+	// height hint. This ensures we don't miss the confirmation even if it
+	// happened around the close height.
+	heightHint -= u.cfg.ConfDepth
```
Comment (yyforyongyu) [inline contractcourt/utxonursery.go:301]: Feel like this should be impossible as the outputs can only be spent after the closing tx is confirmed, but also not very important as we are just scanning a few more blocks.

File: contractcourt/utxonursery.go (hunk 8 above)
Comment (ziggie1984) [reply contractcourt/utxonursery.go:292]: correct

File: contractcourt/utxonursery.go (hunk 9 above)
Comment (ziggie1984) [reply contractcourt/utxonursery.go:301]: took the logic from other parts in the utxo nursery where we would do that, so tried to keep it (maybe there is a race I don't know about haha)

**Comment by litbot-9000** [comment]
@roasbeef: review reminder
@ziggie1984, remember to re-request review from reviewers when ready
